### PR TITLE
feat: Allow more than one action per run

### LINKED ISSUE #2166: Run multiple actions from one invocation
## Run multiple actions from one invocation

Similar to how Gradle might run

```
$ ./gradlew assemble check
```

It would be useful (though low-priority) to run multiple actions in the same invocation of the binary. Perhaps introduce an explicit flag `--action` that can be multi-included? `dagger do --action='foo bar' --action='foo baz'`. This seems preferable than changing the syntax to, perhaps, dot-separated nested actions and whitespace-separated multiple actions (`dagger do foo.bar foo.baz`).

_Originally posted by @NotTheEconomist in https://github.com/dagger/dagger/discussions/2052#discussioncomment-2550137_

### PR DESCRIPTION
**Pull request by NotTheEconomist**
Use the new flag `--action` to specify each action. This change is
backwards-compatible so continuing to use `dagger do actionName` is accepted and
encouraged. In fact they will work together: `dagger do -a action1 action2`
will run both actions.

closes #2166 

Signed-off-by: Adam Smith <adamsmith@nottheeconomist.com>

### CODE REVIEW

**Comment by NotTheEconomist** [comment]
still TBD:
- [ ] tests
- [ ] docs

File: cmd/dagger/cmd/do.go [hunk 1]
```diff
@@ -86,6 +95,28 @@ var doCmd = &cobra.Command{
 	},
 }
 
+func getActions(args []string) ([][]string, error) {
+	actionsFromFlags := viper.GetStringSlice("action")
+	actions := make([][]string, 0, len(actionsFromFlags)+1)
+	for _, flagAction := range actionsFromFlags {
+		action, err := shlex.Split(flagAction)
```
Comment (dolanor) [inline cmd/dagger/cmd/do.go:102]: Do we need shlex?
Wouldn't a simpler `strings.Split(flagAction, " ")` be enough?
What is the use case that would need shell lexing inside of an arg?

Comment (NotTheEconomist) [reply cmd/dagger/cmd/do.go:102]: > Do we need shlex? [...] What is the use case that would need shell lexing?

I actually hadn't considered it, and merely tried to make sure that whatever gets passed to `--action` gets handled the same way as `args` does (lexed by the shell). However it is valid to have an action with a space in its name.

```
"do it": bash.#Run & {
	_image: alpine.#Build & { 
		packages: bash: {}
	}
	input: _image.output
	script: contents: "echo 'its done'"
}
```

`dagger do 'do it'` works here, and `dagger do -a 'do it'` should as well, but wouldn't if we weren't shlexing (I think).

**Comment by helderco** [comment]
This may conflict with https://github.com/dagger/dagger/pull/2033

**Comment by aluzzardi** [comment]
/cc @shykes for DX

This is something I would like as well (be able to run two actions in a same `do`, e.g. `dagger do lint build`).

However, I'm not sure what a good UX could be.

**Comment by NotTheEconomist** [comment]
> This may conflict with #2033

yeah, looks like. I'm open to exploring alternate implementations. In a perfect world we use the DAG here rather than just looping over the actions and evaluating the new target each time. Construct a single plan that forces both targets so shared dependencies are evaluated only once.

> However, I'm not sure what a good UX could be

If we were open to breaking changes then changing `dagger do parent child` to `dagger do parent.child` would allow `dagger do multiple actions at once`. This design was intended to avoid those sorts of breaking changes. It's also possible that this is just not desirable and the solution lies in defining a new action that DOES both those things. Explicit dependencies could get us that, as in make:

```makefile
.PHONY=build-and-lint
build-and-lint: lint build
```

**Comment by helderco** [comment]
But you can already do that, no need to wait for explicit dependencies. Just incorporate other actions inside a new action.

```cue
actions: {
    build: ...
    lint: ...
    buildAndLint: {
        "build": build
        "lint": lint
     }
 }
```

```console
dagger do build
dagger do lint
dagger do buildAndLint
```
```
Or 
```cue
actions: {
    buildAndLint: {
        build: ...
        lint: ...
    }
}
```
```console
dagger do buildAndLint build
dagger do buildAndLint lint
dagger do buildAndLint
```

Choice of names aside 😁

**Comment by NotTheEconomist** [comment]
It looks like there's no clear design on how this should be supported, and building an action like `buildAndLint` above is a workaround for now. Closing until a design is more clear from maintainers.

**Pull request closed without merging**